cumulusci/tasks/bulkdata/load.py

```diff
@@ -142,42 +144,58 @@ def _execute_step(
 
         mapping["oid_as_pk"] = bool(mapping.get("fields", {}).get("Id"))
 
-        bulk_mode = mapping.get("bulk_mode") or self.bulk_mode or "Parallel"
-
-        step = BulkApiDmlOperation(
-            sobject=mapping["sf_object"],
-            operation=(
-                DataOperationType.INSERT
-                if mapping.get("action") == "insert"
-                else DataOperationType.UPDATE
-            ),
-            api_options={"bulk_mode": bulk_mode},
-            context=self,
-            fields=self._get_columns(mapping),
-        )
-
         local_ids = []
+        query = self._query_db(mapping)
+
+        api = mapping.api
+        if api is DataApi.SMART:
+            record_count = query.count()
+            api = DataApi.BULK if record_count > 10000 else DataApi.REST
+
+        if api is DataApi.BULK:
+            bulk_mode = mapping.get("bulk_mode") or self.bulk_mode or "Parallel"
+
+            step = BulkApiDmlOperation(
+                sobject=mapping["sf_object"],
+                operation=(
+                    DataOperationType.INSERT
+                    if mapping.get("action") == "insert"
+                    else DataOperationType.UPDATE
+                ),
+                api_options={"bulk_mode": bulk_mode},
+                context=self,
+                fields=self._get_columns(mapping),
+            )
+        elif api is DataApi.REST:
+            step = RestApiDmlOperation(
+                sobject=mapping["sf_object"],
+                operation=(
+                    DataOperationType.INSERT
+                    if mapping.get("action") == "insert"
+                    else DataOperationType.UPDATE
+                ),
+                api_options={"batch_size": mapping.batch_size},
+                context=self,
+                fields=self._get_columns(mapping),
+            )
```

Comment:
I would guess this logic for picking the API/operation will be easier to test if you refactor it into a separate method.